fgpyo/sam/__init__.py

```diff
@@ -691,6 +691,100 @@ def set_pair_info(r1: AlignedSegment, r2: AlignedSegment, proper_pair: bool = Tr
     r2.template_length = -insert_size
 
 
+def set_mate_info(r1: pysam.AlignedSegment, r2: pysam.AlignedSegment) -> None:
+    """Sets the mate information on a pair of sam records.
+
+    Handles cases where both reads are mapped, one of the two reads is unmapped or both reads
+    are unmapped.
+
+    Args:
+        r1: the first read in the pair
+        r2: the second read in the pair
+    """
+    assert (
+        r1.query_name == r2.query_name
+    ), "Attempting to set mate info on reads with different qnames."
+
+    for rec in r1, r2:
+        rec.template_length = 0
+        rec.is_proper_pair = False
+
+    if r1.is_unmapped and r2.is_unmapped:
+        # If they're both unmapped just clean the records up
+        for rec, other in [(r1, r2), (r2, r1)]:
+            rec.reference_id = NO_REF_INDEX
+            rec.next_reference_id = NO_REF_INDEX
+            rec.reference_start = NO_REF_POS
+            rec.next_reference_start = NO_REF_POS
+            rec.is_unmapped = True
+            rec.is_proper_pair = False
+            rec.mate_is_reverse = other.is_reverse
+            rec.mate_is_unmapped = True
+            rec.set_tag("MC", None)
+
+    elif r1.is_unmapped or r2.is_unmapped:
+        # If only one is mapped/unmapped copy over the relevant stuff
+        (m, u) = (r1, r2) if r2.is_unmapped else (r2, r1)
+        u.reference_id = m.reference_id
+        u.reference_start = m.reference_start
+        u.next_reference_id = m.reference_id
+        u.next_reference_start = m.reference_start
+        u.mate_is_reverse = m.is_reverse
+        u.mate_is_unmapped = False
+        u.set_tag("MC", m.cigarstring)
+
+        m.next_reference_id = u.reference_id
+        m.next_reference_start = u.reference_start
+        m.mate_is_reverse = u.is_reverse
+        m.mate_is_unmapped = True
+        m.set_tag("MC", None)
+
+    else:
+        # Else they are both mapped
+        for rec, other in [(r1, r2), (r2, r1)]:
+            rec.next_reference_id = other.reference_id
+            rec.next_reference_start = other.reference_start
+            rec.mate_is_reverse = other.is_reverse
+            rec.mate_is_unmapped = False
+            rec.set_tag("MC", other.cigarstring)
+
+        if r1.reference_id == r2.reference_id:
+            r1p = r1.reference_end if r1.is_reverse else r1.reference_start
+            r2p = r2.reference_end if r2.is_reverse else r2.reference_start
+            r1.template_length = r2p - r1p
+            r2.template_length = r1p - r2p
+
+            # Arbitrarily set proper pair if the we have an FR pair with isize <= 1000
+            if r1.is_reverse != r2.is_reverse and abs(r1.template_length) <= 1000:
```

Comment:
**issue** This will break many structural variant callers. 

Is there a reason to overwrite the proper pair status reported by the aligner?

Reply:
I'm not sure what initially motivated this; I transferred this `_set_mate_info()` helper function out of `fgpyo/sam/builder.py`. Are you aware of a reason for it to be present there?

Reply:
@msto what's the alternative?  Invoking this function is tantamount to saying "I've changed some of the values on one or both of these reads, please make sure everything is in sync".  I understand the concern around `proper_pair` - and it's true that this implementation made sense when it was in SamBuilder, and less so here...

But if someone has changed alignment values .. the value from the aligner is also questionable here.

Reply:
@coderabbitai what do you think?

Reply:
@TimD1 - @msto and I chatted and we think the right solution here is a little complicated.  We suggest doing something like:

```python
def set_mate_info(
  r1: AlignedSegment,
  r2: AlignedSegment,
  is_proper_pair: Callable[[AlignedSegment, AlignedSegment], bool] = lambda a, b: a.is_proper_pair and b.is_proper_pair
) -> Unit
    ...
```

I.e. the default is to leave it `True` if it's already set on R1 and R1, but set it to false if already false, or there is disagreement.  Then a user (e.g. SamBuilder) can provide it's own lambda.